### PR TITLE
Make `region` optional for S3 instructions

### PR DESCRIPTION
**Pull request by malhotrashivam**
Before this change, `region` was a mandatory parameter for reading parquet files from S3.
After this change, if `region` is not provided, it will be picked up by the AWS sdk using the following provider chain:

```
1. Check the 'aws.region' system property for the region.
2. Check the 'AWS_REGION' environment variable for the region.
3. Check the {user.home}/.aws/credentials and {user.home}/.aws/config files for the region.
4. If running in EC2, check the EC2 metadata service for the region.

Throw exception if not found in any of the locations above
```
Reference:   https://sdk.amazonaws.com/java/api/2.0.0-preview-11/software/amazon/awssdk/awscore/client/builder/AwsClientBuilder.html

Also, since `region` was the only mandatory parameter in S3 instructions, after this change, reading from S3 will be allowed without passing any S3 instructions. For example, code like the following python example will work, as long as the region, access key and secret key are set properly using properties or config files:

```
from deephaven import parquet

test = parquet.read("s3://bucket-name/file-name.parquet")   # Note no s3 instructions
```

### CODE REVIEW

File: extensions/s3/src/main/java/io/deephaven/extensions/s3/S3SeekableChannelProviderPlugin.java [hunk 1]
```diff
@@ -29,9 +29,12 @@ public SeekableChannelsProvider createProvider(@NotNull final URI uri, @Nullable
...
             throw new IllegalArgumentException("Arguments not compatible, provided uri " + uri);
         }
-        if (!(config instanceof S3Instructions)) {
+        if (config != null && !(config instanceof S3Instructions)) {
             throw new IllegalArgumentException("Must provide S3Instructions to read files from S3");
         }
+        if (config == null) {
+            return new S3SeekableChannelProvider(S3Instructions.DEFAULT_S3_INSTRUCTIONS);
+        }
         return new S3SeekableChannelProvider((S3Instructions) config);
```
Comment (malhotrashivam) [inline extensions/s3/src/main/java/io/deephaven/extensions/s3/S3SeekableChannelProviderPlugin.java:38]: Does this particular change look okay to you Devin and Ryan?

File: extensions/s3/src/main/java/io/deephaven/extensions/s3/S3Instructions.java [hunk 2]
```diff
@@ -32,14 +32,19 @@ public abstract class S3Instructions implements LogOutputAppendable {
     private final static Duration DEFAULT_CONNECTION_TIMEOUT = Duration.ofSeconds(2);
     private final static Duration DEFAULT_READ_TIMEOUT = Duration.ofSeconds(2);
 
+    static final S3Instructions DEFAULT_S3_INSTRUCTIONS = builder().build();
```
Comment (lbooker42) [inline extensions/s3/src/main/java/io/deephaven/extensions/s3/S3Instructions.java:35]: I'm not sure how consistent we are in the code base, but I see DEFAULT used often.

```suggestion
    static final S3Instructions DEFAULT = builder().build();
```

File: extensions/s3/src/main/java/io/deephaven/extensions/s3/S3SeekableChannelProviderPlugin.java [hunk 3]
```diff
@@ -29,9 +29,12 @@ public SeekableChannelsProvider createProvider(@NotNull final URI uri, @Nullable
         if (!isCompatible(uri, config)) {
             throw new IllegalArgumentException("Arguments not compatible, provided uri " + uri);
         }
-        if (!(config instanceof S3Instructions)) {
+        if (config != null && !(config instanceof S3Instructions)) {
```
Comment (lbooker42) [inline extensions/s3/src/main/java/io/deephaven/extensions/s3/S3SeekableChannelProviderPlugin.java:32]: Is this the correct error message? Suggestion: "Only S3Instructions are valid when reading files from S3"

**Review by rcaudy — COMMENTED** [review]
My comments are the same as Larry's.

**Review by rcaudy — COMMENTED** [review]
Caught up

File: extensions/s3/src/main/java/io/deephaven/extensions/s3/S3SeekableChannelProviderPlugin.java [hunk 4]
```diff
@@ -29,9 +29,12 @@ public SeekableChannelsProvider createProvider(@NotNull final URI uri, @Nullable
...
             throw new IllegalArgumentException("Must provide S3Instructions to read files from S3");
         }
+        if (config == null) {
+            return new S3SeekableChannelProvider(S3Instructions.DEFAULT_S3_INSTRUCTIONS);
+        }
```
[the remark this answers is not in the document]
Comment (malhotrashivam) [reply extensions/s3/src/main/java/io/deephaven/extensions/s3/S3SeekableChannelProviderPlugin.java]: Discussed offline.

File: py/server/tests/test_parquet.py [hunk 5]
```diff
@@ -570,19 +570,16 @@ def test_read_parquet_from_s3(self):
...
         # Fails since we provide the key without the secret key
         with self.assertRaises(DHError):
             s3.S3Instructions(region_name="us-east-1",
                               access_key_id="Some key without secret",
                               )
 
-        s3_instructions = s3.S3Instructions(region_name="us-east-1",
-                                            read_ahead_count=1,
-                                            )
+        s3_instructions = s3.S3Instructions()
```
Comment (devinrsmith) [inline py/server/tests/test_parquet.py:581]: We should probably have a test where we don't pass `special_instructions` in too?

**Review by devinrsmith — APPROVED** [review]

**Review by jmao-denver — APPROVED** [review]
Python changes LGTM

**Comment by deephaven-internal** [comment]
Labels indicate documentation is required. Issues for documentation have been opened:

Community: https://github.com/deephaven/deephaven-docs-community/issues/234

**Pull request merged**